persistence/src/commonMain/kotlin/com/quran/shared/persistence/repository/BookmarksRepositoryImpl.kt

```diff
@@ -0,0 +1,245 @@
+import com.quran.shared.persistence.QuranDatabase
+import com.quran.shared.persistence.model.Bookmark
+import com.quran.shared.persistence.model.BookmarkMutation
+import com.quran.shared.persistence.model.BookmarkMutationType
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.IO
+
+class BookmarksRepositoryImpl(
+    private val database: QuranDatabase
+) : BookmarksRepository, BookmarksSynchronizationRepository {
+    private val logger = Logger.withTag("BookmarksRepository")
+
+    override fun getAllBookmarks(): Flow<List<Bookmark>> {
+        val persistedFlow = database.bookmarksQueries.getBookmarks()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toBookmark() }
+            }
+        val mutatedFlow = database.bookmarks_mutationsQueries.getBookmarksMutations()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toBookmarkMutation() }
+            }
+
+        return merge(persistedFlow, mutatedFlow) { it.toBookmark() }
+    }
+
+    override fun getPageBookmarks(): Flow<List<Bookmark.PageBookmark>> {
+        // sqldelight created different return types for these queries, so we couldn't
+        // share the full logic with the other fetching methods.
+        val persistedFlow = database.bookmarksQueries.getPageBookmarks()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toPageBookmark() }
+            }
+        val mutatedFlow = database.bookmarks_mutationsQueries.getPageBookmarkMutations()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toBookmarkMutation() }
+            }
+
+        return merge(persistedFlow, mutatedFlow) { it.toPageBookmark() }
+    }
+
+    override fun getAyahBookmarks(): Flow<List<Bookmark.AyahBookmark>> {
+        val persistedFlow = database.bookmarksQueries.getAyahBookmarks()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toAyahBookmark() }
+            }
+        val mutatedFlow = database.bookmarks_mutationsQueries.getAyahBookmarkMutations()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toBookmarkMutation() }
+            }
+
+        return merge(persistedFlow, mutatedFlow) { it.toAyahBookmark() }
+    }
+
+    private fun <T: Bookmark>merge(
+        persisted: Flow<List<T>>,
+        mutated: Flow<List<BookmarkMutation>>,
+        mapper: (BookmarkMutation) -> T?
+    ): Flow<List<T>> = persisted.combine(mutated) { persistedBookmarks, mutatedBookmarks ->
+        val deletedRemoteIDs = mutatedBookmarks
+            .filter { it.remoteId != null } // TODO: This is confusing, but it's equivalent to the required result.
```

Comment:
the comment above will help you get rid of this since you can just make it non-null for those cases